prometheus/client.go

```diff
@@ -33,40 +31,72 @@ func NewClient() (*Client, error) {
 	if err != nil {
 		return nil, err
 	}
-	client.p8s = p8s
+	client.api = v1.NewAPI(p8s)
 	return &client, nil
 }
 
+// For testing / mocking
+func (in *Client) inject(api v1.API) {
+	in.api = api
+}
+
 // GetSourceServices returns a map of source services for a given service identified by its namespace and service name.
 // Returned map has a destination version as a key and a "<origin service>/<origin version>" pair as value.
 // Destination service is not included in the map as it is passed as argument.
 // It returns an error on any problem.
-func (in *Client) GetSourceServices(namespace string, servicename string) (map[string]string, error) {
-	query := fmt.Sprintf(istioRequestQuery, servicename, namespace)
-	api := v1.NewAPI(in.p8s)
-	result, err := api.Query(context.Background(), query, time.Now())
+func (in *Client) GetSourceServices(namespace string, servicename string) (map[string][]string, error) {
+	query := fmt.Sprintf("istio_request_count{destination_service=\"%s.%s.%s\"}",
+		servicename, namespace, config.Get().IstioIdentityDomain)
+	result, err := in.api.Query(context.Background(), query, time.Now())
 	if err != nil {
 		return nil, err
 	}
-	routes := make(map[string]string)
+	routes := make(map[string][]string)
 	switch result.Type() {
 	case model.ValVector:
 		matrix := result.(model.Vector)
 		for _, sample := range matrix {
 			metric := sample.Metric
 			index := fmt.Sprintf("%s", metric["destination_version"])
 			sourceService := string(metric["source_service"])
-			// .svc sufix is a pure Istio label, I guess we can skip it at the moment for clarity
-			if i := strings.Index(sourceService, ".svc"); i > 0 {
+			// sourceService is in the form "service.namespace.istio_identity_domain". We want to keep only "service.namespace".
+			if i := strings.Index(sourceService, "."+config.Get().IstioIdentityDomain); i > 0 {
 				sourceService = sourceService[:i]
 			}
-			routes[index] = fmt.Sprintf("%s/%s", sourceService, metric["source_version"])
+			source := fmt.Sprintf("%s/%s", sourceService, metric["source_version"])
```

Comment:
+1
These change allows to cover cases where we might have
```
ProductServices
Destination    Sources
v1                   Review/v1, Review/v2
v2                   Review/v3, Review/v4
```
So before we cover only
```
Destination    Sources
v1                   Review/v1
v2                   Review/v3
```
Which misses potential versions for a single destination.
Good change.